format/ComputeIR.fbs

```diff
@@ -0,0 +1,241 @@
+  options: Buffer;
+
+  /// The arguments passed to `relation_name`.
+  arguments: [Relation] (required);
+
+  /// The schema of rows in this Relation
+  schema: Schema;
+}
+
+/// The contents of Relation.options will be FilterOptions
+/// if Relation.relation_name = "filter"
+table FilterOptions {
+  /// The expression which will be evaluated against input rows
+  /// to determine whether they should be excluded from the
+  /// "filter" relation's output.
+  filter_expression: Expression (required);
+}
+
+/// The contents of Relation.options will be ProjectOptions
+/// if Relation.relation_name = "project"
+table ProjectOptions {
+  /// Expressions which will be evaluated to produce to
+  /// the rows of the "project" relation's output.
+  expressions: [Expression] (required);
+}
+
+/// The contents of Relation.options will be AggregateOptions
+/// if Relation.relation_name = "aggregate"
+table AggregateOptions {
+  /// Expressions which will be evaluated to produce to
+  /// the rows of the "aggregate" relation's output.
+  aggregations: [Expression] (required);
+  /// Keys by which `aggregations` will be grouped.
+  keys: [Expression];
+}
+
+/// The contents of Relation.options will be JoinOptions
+/// if Relation.relation_name = "join"
+table JoinOptions {
+  /// The expression which will be evaluated against rows from each
+  /// input to determine whether they should be included in the
+  /// "join" relation's output.
+  on_expression: Expression (required);
+  join_kind: string;
+}
+
+/// Whether lesser values should precede greater or vice versa,
+/// also whether nulls should preced or follow values.
+enum Ordering : uint8 {
+  ASCENDING_THEN_NULLS,
+  DESCENDING_THEN_NULLS,
+  NULLS_THEN_ASCENDING,
+  NULLS_THEN_DESCENDING
+}
+
+table SortKey {
+  value: Expression (required);
+  ordering: Ordering = ASCENDING_THEN_NULLS;
+}
+
+/// The contents of Relation.options will be OrderByOptions
+/// if Relation.relation_name = "order_by"
+table OrderByOptions {
+  /// Define sort order for rows of output.
+  /// Keys with higher precedence are ordered ahead of other keys.
+  keys: [SortKey] (required);
+}
+
+/// The contents of Relation.options will be LimitOptions
+/// if Relation.relation_name = "limit"
+table LimitOptions {
+  /// Set the maximum number of rows of output.
+  count: long;
+}
+
+/// The contents of Relation.options will be CommonOptions
+/// if Relation.relation_name = "common"
+table CommonOptions {
+  /// Commons (CTEs in SQL) allow assigning a name to a stream
+  /// of data and reusing it, potentially multiple times and
+  /// potentially recursively.
+  name: string;
+}
+
+/// The contents of Relation.options will be UnionOptions
+/// if Relation.relation_name = "union"
+table UnionOptions {
+  /// For simplicity, all rows from any input to a "union" relation
+  /// will always be concatenated into a single output- establishing
+  /// uniqueness of output rows is deferred to other relations.
```

Comment:
I'm not entirely sure we can leave out something like a `unique: boolean;` parameter, since chaining unions with different value of such a parameter is (syntactically) legal:

```sql
SELECT 1
UNION ALL
SELECT 1
UNION
SELECT 1
```

Maybe there's an algebraic equivalence here that obviates the need for such a parameter, but I don't see it.

Reply:
Wouldn't that be equivalent to
```sql
with s as (
  select 1
  union all
  select 1
  union all
  select 1
)
select distinct * from s;
```
?

Reply:
It is, but I'm wondering if there are cases where a union doesn't collapse into a `select distinct *` query.

Reply:
You can always rewrite a single `select … UNION select…` as `select distinct * from (select … UNION ALL select …)` so I would think they're always equivalent.

Reply:
Thanks for the response!